File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/safe-turns.ts
/**
 * Campaign-scoped safe turn submission. Each campaign tracks a single
 * `current_turn` counter starting at 1. A submission is only accepted when
 * its `expected_turn` matches the current counter, which is what makes
 * concurrent/racing submissions for a stale turn safe to reject without
 * mutating queue state.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, isMember } from './shared.ts';

interface SafeTurnStateRow {
  current_turn: number;
}

interface SafeTurnSubmissionRow {
  sequence: number;
  submission_id: string;
  action: string;
  accepted_turn: number;
  next_turn: number;
}

function getOrInitState(db: ReturnType<typeof getDb>, campaignId: string): number {
  const row = db
    .prepare('SELECT current_turn FROM play_campaign_safe_turns WHERE campaign_id = ?')
    .get(campaignId) as SafeTurnStateRow | undefined;
  if (row) return row.current_turn;

  db.prepare('INSERT INTO play_campaign_safe_turns (campaign_id, current_turn) VALUES (?, 1)').run(campaignId);
  return 1;
}

function toSubmissionBody(row: SafeTurnSubmissionRow): JsonValue {
  return {
    submission_id: row.submission_id,
    action: row.action,
    accepted_turn: row.accepted_turn,
    next_turn: row.next_turn,
  } as unknown as JsonValue;
}

export function submitSafeTurn(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const isOwner = actor.username === campaign.owner;
  if (!isOwner && !isMember(db, campaignId, actor.username)) {
    return { status: 403, body: { error: 'only campaign members may submit safe turns' } };
  }

  const submissionId = body.submission_id;
  if (typeof submissionId !== 'string' || submissionId.length === 0) {
    return { status: 400, body: { error: 'submission_id must be a non-empty string' } };
  }

  const action = body.action;
  if (typeof action !== 'string' || action.length === 0) {
    return { status: 400, body: { error: 'action must be a non-empty string' } };
  }

  const expectedTurn = body.expected_turn;
  if (typeof expectedTurn !== 'number' || !Number.isInteger(expectedTurn) || expectedTurn <= 0) {
    return { status: 400, body: { error: 'expected_turn must be a positive integer' } };
  }

  const currentTurn = getOrInitState(db, campaignId);

  const existingSubmission = db
    .prepare('SELECT sequence FROM play_campaign_safe_turn_submissions WHERE campaign_id = ? AND submission_id = ?')
    .get(campaignId, submissionId);
  if (existingSubmission) {
    return { status: 409, body: { error: 'submission_id already used in this campaign' } };
  }

  if (expectedTurn !== currentTurn) {
    return { status: 409, body: { current_turn: currentTurn } as unknown as JsonValue };
  }

  const nextTurn = currentTurn + 1;
  const sequenceRow = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_safe_turn_submissions WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };
  const sequence = sequenceRow.max_sequence + 1;

  db.prepare(
    'INSERT INTO play_campaign_safe_turn_submissions (campaign_id, sequence, submission_id, action, accepted_turn, next_turn) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequence, submissionId, action, currentTurn, nextTurn);

  db.prepare('UPDATE play_campaign_safe_turns SET current_turn = ? WHERE campaign_id = ?').run(nextTurn, campaignId);

  return {
    status: 201,
    body: { submission_id: submissionId, action, accepted_turn: currentTurn, next_turn: nextTurn } as unknown as JsonValue,
  };
}

export function getSafeTurns(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const isOwner = actor.username === campaign.owner;
  if (!isOwner && !isMember(db, campaignId, actor.username)) {
    return { status: 403, body: { error: 'only the campaign DM or members may read safe turns' } };
  }

  const currentTurn = getOrInitState(db, campaignId);

  const rows = db
    .prepare(
      'SELECT sequence, submission_id, action, accepted_turn, next_turn FROM play_campaign_safe_turn_submissions WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as unknown as SafeTurnSubmissionRow[];

  return {
    status: 200,
    body: { current_turn: currentTurn, accepted: rows.map(toSubmissionBody) } as unknown as JsonValue,
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/safe_turns.py
"""Campaign-scoped safe turn submission views."""

import sqlite3

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from .common import require_play_campaign_member_or_owner
from ..http import bad_request, conflict, parse_json, require_method
from ..db import db_conn


@csrf_exempt
def safe_turns(request, id):
    """Dispatch GET/POST for /v1/play/campaigns/{id}/safe-turns."""
    bad = require_method(request, "GET", "POST")
    if bad:
        return bad
    if request.method == "POST":
        return submit_safe_turn(request, id)
    return get_safe_turns(request, id)


@csrf_exempt
def submit_safe_turn(request, id):
    """Accept or reject a deterministic safe-turn submission."""
    bad = require_method(request, "POST")
    if bad:
        return bad

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")
    try:
        submission_id = body["submission_id"]
        expected_turn = body["expected_turn"]
        action = body["action"]
        if not isinstance(submission_id, str) or not submission_id:
            raise ValueError
        if not isinstance(action, str) or not action:
            raise ValueError
        if not isinstance(expected_turn, int) or isinstance(expected_turn, bool) or expected_turn < 1:
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        current_turn = campaign["safe_turn_current_turn"]

        existing = conn.execute(
            "SELECT 1 FROM safe_turn_submissions WHERE campaign_id = ? AND submission_id = ?",
            (id, submission_id),
        ).fetchone()
        if existing is not None:
            return conflict("duplicate submission_id")

        if expected_turn != current_turn:
            conn.execute(
                "INSERT INTO safe_turn_submissions (campaign_id, submission_id, action, accepted_turn, next_turn, accepted) "
                "VALUES (?, ?, ?, NULL, NULL, 0)",
                (id, submission_id, action),
            )
            return JsonResponse({"current_turn": current_turn}, status=409)

        next_turn = current_turn + 1
        conn.execute(
            "INSERT INTO safe_turn_submissions (campaign_id, submission_id, action, accepted_turn, next_turn, accepted) "
            "VALUES (?, ?, ?, ?, ?, 1)",
            (id, submission_id, action, current_turn, next_turn),
        )
        conn.execute(
            "UPDATE play_campaigns SET safe_turn_current_turn = ? WHERE id = ?",
            (next_turn, id),
        )

    return JsonResponse(
        {
            "submission_id": submission_id,
            "action": action,
            "accepted_turn": current_turn,
            "next_turn": next_turn,
        },
        status=201,
    )


@csrf_exempt
def get_safe_turns(request, id):
    """Return the ordered accepted safe-turn history and current turn."""
    bad = require_method(request, "GET")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        rows = conn.execute(
            "SELECT submission_id, action, accepted_turn, next_turn FROM safe_turn_submissions "
            "WHERE campaign_id = ? AND accepted = 1 ORDER BY id ASC",
            (id,),
        ).fetchall()

    accepted = [
        {
            "submission_id": row["submission_id"],
            "action": row["action"],
            "accepted_turn": row["accepted_turn"],
            "next_turn": row["next_turn"],
        }
        for row in rows
    ]

    return JsonResponse(
        {
            "current_turn": campaign["safe_turn_current_turn"],
            "accepted": accepted,
        }
    )
